pkg/mads/v1/service/service.go

```diff
@@ -1,46 +1,65 @@
 package service
 
 import (
+	"context"
+	"time"
+
+	envoy_cache "github.com/envoyproxy/go-control-plane/pkg/cache/v3"
 	"github.com/go-logr/logr"
 
 	"github.com/kumahq/kuma/pkg/config/mads"
 	core_manager "github.com/kumahq/kuma/pkg/core/resources/manager"
 	mads_generator "github.com/kumahq/kuma/pkg/mads/v1/generator"
 	mads_reconcile "github.com/kumahq/kuma/pkg/mads/v1/reconcile"
+	util_watchdog "github.com/kumahq/kuma/pkg/util/watchdog"
 	util_xds "github.com/kumahq/kuma/pkg/util/xds"
 	util_xds_v3 "github.com/kumahq/kuma/pkg/util/xds/v3"
 	"github.com/kumahq/kuma/pkg/xds/cache/mesh"
 )
 
 type service struct {
-	server Server
-	config *mads.MonitoringAssignmentServerConfig
-	log    logr.Logger
+	server   Server
+	config   *mads.MonitoringAssignmentServerConfig
+	log      logr.Logger
+	watchdog *util_watchdog.SimpleWatchdog
 }
 
 func NewService(config *mads.MonitoringAssignmentServerConfig, rm core_manager.ReadOnlyResourceManager, log logr.Logger, meshCache *mesh.Cache) *service {
-	hasher := util_xds_v3.IDHash{}
-	cache := util_xds_v3.NewSnapshotCache(false, hasher, util_xds.NewLogger(log))
+	hasher := &util_xds_v3.FallBackNodeHash{DefaultId: mads_generator.DefaultKumaClientId}
+	cache := envoy_cache.NewSnapshotCache(false, hasher, util_xds.NewLogger(log))
 	reconciler := mads_reconcile.NewReconciler(
-		hasher,
 		cache,
 		mads_reconcile.NewSnapshotGenerator(rm, mads_generator.MonitoringAssignmentsGenerator{}, meshCache),
 	)
-
-	callbacks := util_xds_v3.CallbacksChain{
-		util_xds_v3.AdaptMultiCallbacks(util_xds.LoggingCallbacks{Log: log}),
-		// TODO Right now we are creating sync tracker for first REST request pkg/util/xds/v3/watchdog_callbacks.go:48, because of this
-		// using MeshMetrics with specified clientId will refresh only one client in background
-		// issue: https://github.com/kumahq/kuma/issues/8764
-		NewSyncTracker(reconciler, config.AssignmentRefreshInterval.Duration, log),
-		// For on-demand reconciliation
-		util_xds_v3.AdaptRestCallbacks(NewReconcilerRestCallbacks(reconciler)),
+	// We use the clientIds from the reconciler from the node hasher
+	hasher.GetIds = reconciler.KnownClientIds
+	watchdog := &util_watchdog.SimpleWatchdog{
+		NewTicker: func() *time.Ticker {
+			return time.NewTicker(config.AssignmentRefreshInterval.Duration)
+		},
+		OnTick: func(ctx context.Context) error {
+			log.V(1).Info("on tick")
+			err := reconciler.Reconcile(ctx)
+			return err
+		},
+		OnError: func(err error) {
+			log.Error(err, "OnTick() failed")
+		},
 	}
-	srv := NewServer(cache, callbacks)
+	watchdog.WithTickCheck()
+	srv := NewServer(cache, util_xds_v3.CallbacksChain{
+		util_xds_v3.AdaptMultiCallbacks(util_xds.LoggingCallbacks{Log: log}),
+	})
 
 	return &service{
-		server: srv,
-		config: config,
-		log:    log,
+		server:   srv,
+		config:   config,
+		log:      log,
+		watchdog: watchdog,
 	}
 }
+
+func (s *service) Start(ctx context.Context) {
+	go s.watchdog.Start(ctx)
+	s.watchdog.HasTicked(true)
```

Comment:
couldn't you just call `s.watchdog.OnTick()` instead of those tick checks?

Reply:
No because `OnTick` is rarely threadSafe, I find `HasTicked` something that could be quite useful anyway as a construct...

Reply:
I meant before starting the watchdog, then the thread safety is not a concern